### ISSUE TITLE
[BUG] auto update fails after dynaconf refactor

### ISSUE THREAD

**Issue by ivelin** [ambianic/ambianic-edge#224]
**Describe the bug**

watchtower updated to the new dynaconf edge version and it now fails due to inability to read a previous config.yaml format. We should push a new version that gracefully converts a previous config.yaml to the new version without bricking edge.

**To Reproduce**
1. Get an existing ambianic edge config.yaml which uses yaml entity references
2. Download and run new docker image with dynaconf patch
3. Look at  docker-compose logs

```
ambianic-edge    | 2020-11-17 22:21:12 INFO /workspace/peerjs/ext/http-proxy.py.peer_open(91): Peer signaling connection open.
ambianic-edge    | 2020-11-17 22:21:12 INFO /workspace/peerjs/ext/http-proxy.py.peer_open(105): savedPeerId: 35bdf700-cbff-4358-8a0e-d1ee70ceef67
ambianic-edge    | Traceback (most recent call last):
ambianic-edge    |   File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
ambianic-edge    |     "__main__", mod_spec)
ambianic-edge    |   File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
ambianic-edge    |     exec(code, run_globals)
ambianic-edge    |   File "/opt/ambianic-edge/src/ambianic/__main__.py", line 46, in <module>
ambianic-edge    |     main()
ambianic-edge    |   File "/opt/ambianic-edge/src/ambianic/__main__.py", line 19, in main
ambianic-edge    |     ambianic.server_instance.start()
ambianic-edge    |   File "/opt/ambianic-edge/src/ambianic/server.py", line 129, in start
ambianic-edge    |     logger.configure(config.get("logging"))
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/base.py", line 141, in __getattr__
ambianic-edge    |     self._setup()
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/base.py", line 192, in _setup
ambianic-edge    |     settings_module=settings_module, **self._kwargs
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/base.py", line 255, in __init__
ambianic-edge    |     self.execute_loaders()
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/base.py", line 948, in execute_loaders
ambianic-edge    |     self, env=env, silent=silent, key=key, filename=filename
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/loaders/__init__.py", line 125, in settings_loader
ambianic-edge    |     obj, filename=mod_file, env=env, silent=silent, key=key
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/loaders/yaml_loader.py", line 52, in load
ambianic-edge    |     loader.load(filename=filename, key=key, silent=silent)
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/loaders/base.py", line 62, in load
ambianic-edge    |     source_data = self.get_source_data(files)
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/loaders/base.py", line 83, in get_source_data
ambianic-edge    |     content = self.file_reader(open_file)
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/vendor/ruamel/yaml/main.py", line 341, in safe_load
ambianic-edge    |     def safe_load(stream,version=_A):return load(stream,SafeLoader,version)
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/vendor/ruamel/yaml/main.py", line 323, in load
ambianic-edge    |     try:return loader._constructor.get_single_data()
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/vendor/ruamel/yaml/constructor.py", line 107, in get_single_data
ambianic-edge    |     node=self.composer.get_single_node()
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/vendor/ruamel/yaml/composer.py", line 34, in get_single_node
ambianic-edge    |     if not A.parser.check_event(StreamEndEvent):B=A.compose_document()
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/vendor/ruamel/yaml/composer.py", line 37, in compose_document
ambianic-edge    |     def compose_document(A):A.parser.get_event();B=A.compose_node(_A,_A);A.parser.get_event();A.anchors={};return B
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/vendor/ruamel/yaml/composer.py", line 49, in compose_node
ambianic-edge    |     elif A.parser.check_event(MappingStartEvent):E=A.compose_mapping_node(B)
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/vendor/ruamel/yaml/composer.py", line 74, in compose_mapping_node
ambianic-edge    |     while not B.parser.check_event(MappingEndEvent):G=B.compose_node(A,_A);H=B.compose_node(A,G);A.value.append((G,H))
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/vendor/ruamel/yaml/composer.py", line 49, in compose_node
ambianic-edge    |     elif A.parser.check_event(MappingStartEvent):E=A.compose_mapping_node(B)
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/vendor/ruamel/yaml/composer.py", line 74, in compose_mapping_node
ambianic-edge    |     while not B.parser.check_event(MappingEndEvent):G=B.compose_node(A,_A);H=B.compose_node(A,G);A.value.append((G,H))
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/vendor/ruamel/yaml/composer.py", line 49, in compose_node
ambianic-edge    |     elif A.parser.check_event(MappingStartEvent):E=A.compose_mapping_node(B)
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/vendor/ruamel/yaml/composer.py", line 74, in compose_mapping_node
ambianic-edge    |     while not B.parser.check_event(MappingEndEvent):G=B.compose_node(A,_A);H=B.compose_node(A,G);A.value.append((G,H))
ambianic-edge    |   File "/usr/local/lib/python3.7/dist-packages/dynaconf/vendor/ruamel/yaml/composer.py", line 41, in compose_node
ambianic-edge    |     if D not in A.anchors:raise ComposerError(_A,_A,'found undefined alias %r'%utf8(D),C.start_mark)
ambianic-edge    | dynaconf.vendor.ruamel.yaml.composer.ComposerError: found undefined alias 'secret_uri_front_door_camera'
ambianic-edge    |   in "/workspace/config.yaml", line 22, column 10
ambianic-edge exited with code 1
ambianic-edge    | Reading package lists...
```

**Expected behavior**

Clean upgrade that's transparent to the user. Read config.yaml with a standard yaml library and convert it to a simplified format that dynaconf understands. Then save the new version back to config.yaml. Log an INFO message letting the user know about the conversion with a link to online document that explains the new format.

**Host environment (please complete the following information):**
 - Hardware: Raspberry Pi
 - OS: Raspbian OS Buster
 - Container docker, docker-compose, watchtower
 - Version: Ambianic-edge v1.8.5 
https://github.com/ambianic/ambianic-edge/releases/tag/v1.8.5

**Additional context**

Related to #222

**Comment by ivelin** [comment]
Update: I was able to manually fix the problem by merging the contents of  secrets.yaml into config.yaml. 
This is not something a regular user would know how to do without digging into the code. There is nothing in the PWA UI that suggests what the root cause of the problem was. It just shows that the UI cannot connect to the edge device.

We should be cautious about disruptive upgrades to avoid user perceived downtime.

**Comment by muka** [comment]
Thanks for reporting, the best way would be to move this logic in the quickstart repository where there is already an upgrade routine that we can extend with dedicated merge and update logic

**Comment by muka** [comment]
Can you share your config and secrets? I suspect that the upgrade required is mostly related to remove yaml pointers. Is it the case? Thanks!

**Comment by ivelin** [comment]
The yaml pointers actually worked once I moved the entities from
secrets.yaml to config.yaml. Other yaml pointers within config.yaml worked
fine.

However the code that used to prepend secrets.yaml (if it exists) to
config.yaml before passing the resulting file to an yaml parser is now gone
and config.yaml's pointers cannot be resolved.

**Comment by ivelin** [comment]
Before the upgrade, I used to have a few entities like this one in my secrets.yaml:

```
secret_uri_front_door_camera: &secret_uri_front_door_camera 'http://192.168.86.131/cgi-bin/snapshot.cgi?chn=0&.....
```

Which were then referenced in config.yaml as follows:

```
sources:
  front_door_camera: &src_front_door_cam
    uri: *secret_uri_front_door_camera
    # type: video, audio, or auto
    type: image
    live: true

```


After the upgrade failed, I manually moved the entities from secrets.yaml to config.yaml and the upgraded ambianic-edge image worked without the errors I shared above.

**Comment by ivelin** [comment]
Looks liky dynaconf supports a very similar approach, so we should probably find a way to accommodate the existing secrets.yaml somehow:

https://dynaconf.readthedocs.io/en/2.2.3/guides/sensitive_secrets.html

> Using .secrets files
> To safely store sensitive data Dynaconf also searches for a .secrets.{toml|py|json|ini|yaml} file to look for data like tokens and passwords.

> example .secrets.toml:
```
[default]
password = "sek@987342$"
```
> The secrets file supports all the environment definitions supported in the settings file.

> IMPORTANT: The reason to use a .secrets.* file is the ability to omit this file when committing to the repository so a recommended .gitignore should include .secrets.* line.

**Comment by muka** [comment]
That's a good thing, code already handles loading the secrets file yaml but has been commented out. We just need to ensure those configurations are actually merged by dynaconf and we can have back the same feature.

Looking from a component perspective, I would move datasources to a dedicated service which cold expose an for API configuration and also to access stream data from a common interface.
With this we could avoid setting such config in a static file but expose in the UI for example.

**Comment by ivelin** [comment]
> That's a good thing, code already handles loading the secrets file yaml but has been commented out. We just need to ensure those configurations are actually merged by dynaconf and we can have back the same feature.

Nice. 

> Looking from a component perspective, I would move datasources to a dedicated service which cold expose an for API configuration and also to access stream data from a common interface. With this we could avoid setting such config in a static file but expose in the UI for example.

That makes sense to me as a project direction.

**Comment by muka** [comment]
I've investigated how to bring back secrets. Cross yaml pointer won't work with the new implementation. The proposed solution is to use a placeholder such as `${key}`.  

An example usage

secrets.yaml
```yaml
src_front_door_cam: 'http://192.168.86.131/cgi-bin/snapshot.cgi?chn=0'
```

config.yaml
```yaml
sources:
  front_door_camera: 
    uri: ${secret_uri_front_door_camera}
    # type: video, audio, or auto
    type: image
    live: true
```

In case the placeholder is not found, the key is not replaced. Additionally, this solution works also with other formats than yaml.
Implementation is available in #251
